storage/merkle/proof.go

```diff
@@ -6,80 +6,97 @@ import (
+// Proof captures all data needed for inclusion proof of a single value inserted under key `Key` into the merkle trie
 type Proof struct {
 	// key that is used to insert and look up the value
 	Key []byte
 	// value stored in the trie for the given key
 	Value []byte
-	// size of this is equal to the steps needed for verification,
-	// constructed by traversing top to down, if set to zero means we had a full node and a value from a hash value from InterimHashes should be read,
-	// any other value than 0 means we hit a short node and we need the shortcounts for computing the hash
+
+	// InterimNodeTypes encodes the type of each visited node on the path. We use the following convention:
+	//  * Let `t := InterimNodeTypes[i]`. Index `i` represents node along the path with the distance
+	//    `i` (number of edges) from the trie root.
+	//  * The value `t` represents the node's type:
+	//    • `t == 0` indicates that the vertex is a full node
+	//    • `t == k`, for any `k > 0`, indicates that vertex is a short node `s`, with `s.count == k`
 	//
 	// TODO: as this would include many zeros and would be sparse if the keys are computed as output of hashes, we might optimize
 	// this by using an sparse representation in the future (index and counts for non-zeros only)
-	ShortCounts []int
-	// a slice of hash values (hash value of siblings for full nodes, the size of this is equal to the zeros we have in ShortCounts)
-	InterimHashes [][]byte
+	InterimNodeTypes []uint32
+
+	// SiblingHashes hold the hash of the non-visited sibling node for each branch on the path.
+	// Elements are ordered from root to leaf. As branches are represented as full nodes, the
+	// size of this slice must be equal to the zeros we have in InterimNodeTypes.
+	SiblingHashes [][]byte
 }
 
 // Verify verifies the proof by constructing the hash values bottom up and cross check
 // the constructed root hash with the given one.
 // if the proof is valid it returns true and false otherwise
 func (p *Proof) Verify(expectedRootHash []byte) bool {
-	// iterate backward and verify the proof
-	currentHash := computeLeafHash(p.Value)
-	hashIndex := len(p.InterimHashes) - 1
-
-	// compute last path index
-	pathIndex := len(p.InterimHashes)
-	for _, sc := range p.ShortCounts {
-		pathIndex += sc
+	// before we verify any hashes, check correct formatting of the proof
+	if len(p.Key) < 1 || maxKeyLength < len(p.Key) {
+		return false
+	}
+	expectedKeyLength := uint32(0)
+	expectedNumberSiblingHashes := 0
+	for _, t := range p.InterimNodeTypes {
+		if t == 0 { // full node
+			expectedKeyLength += 1
+			expectedNumberSiblingHashes += 1
+		} else { // short node
+			expectedKeyLength += t
+		}
+		if maxKeyLength < expectedKeyLength { // preventing overflow attack
+			return false
+		}
 	}
+	if expectedKeyLength != uint32(len(p.Key)) || expectedNumberSiblingHashes != len(p.SiblingHashes) {
```

Comment:
sorry, forgot the bite -> bit conversion here:
```suggestion
	if expectedKeyLength != uint32(len(p.Key) * 8) || expectedNumberSiblingHashes != len(p.SiblingHashes) {
```